Refactoring to create a 'wallet' package

manan19 opened the pull request.

jrick commented:
Please don't add a global wallet instance to the main package.  You can end up with races between reading and setting the pointer.  Instead, use the wallet pointers that are passed down to all of the handlers, after the wallet has been loaded and added to the server with `SetWallet`.

jrick commented on btcwallet.go:
```diff
@@ -25,11 +25,13 @@ import (
 	"runtime"
 
 	"github.com/btcsuite/btcwallet/chain"
+	"github.com/mirrorx/btcwallet/wallet"
```
btcsuite

manan19 replied:
I'll change this at the very end - right before it's ready to merge.

jrick commented on config.go:
```diff
@@ -24,24 +24,23 @@ import (
 	defaultLogDirname        = "logs"
 	defaultLogFilename       = "btcwallet.log"
 	defaultDisallowFree      = false
+	defaultUseAddrIndex      = false
```
This should be in a different PR

manan19 replied:
removed it.

jrick commented on config.go:
```diff
@@ -49,6 +48,7 @@ var (
 	btcwalletHomeDir   = btcutil.AppDataDir("btcwallet", false)
 	btcdHomedirCAFile  = filepath.Join(btcdHomeDir, "rpc.cert")
 	defaultConfigFile  = filepath.Join(btcwalletHomeDir, defaultConfigFilename)
+	defaultPubPassphrase = "public"
```
This should probably be exported from another package.

Also, this file is not `go fmt`-formatted.

manan19 replied:
Could you please elaborate? Do you want defaultPubPassphrase only to be exported from another package? 

go fmt - addressed.

jrick replied:
It doesn't belong in the main package.  I would add it to waddrmgr and export it so the wallet package can use it.

manan19 commented:
This removes the global wallet ref
https://github.com/btcsuite/btcwallet/commit/87c18979ef4b4e2ca39c535f6086a2820f53c7ee

jrick commented on config.go:
```diff
@@ -263,7 +255,7 @@ func loadConfig() (*config, []string, error) {
 	// Pre-parse the command line options to see if an alternative config
 	// file or the version flag was specified.
 	preCfg := cfg
```
This has changed semantics now because cfg is a pointer, not a `config`, so it doesn't make a shallow copy of the struct.  Is this intentional?

manan19 replied:
changed it back to how it was originally

jrick commented on log.go:
```diff
@@ -42,33 +34,19 @@ const (
 // subsystemLoggers maps each subsystem identifier to its associated logger.
 var subsystemLoggers = map[string]btclog.Logger{
-	"BTCW": log,
+	"BTCWD": log,
```
Try to keep all logging subsystems 4 chars long

manan19 replied:
addressed

jrick commented on wallet/config.go:
```diff
@@ -0,0 +1,38 @@
+	WalletDbName     string
+}
+
+func (w *Wallet) GetConfig() (*Config) {
```
I would prefer if the config was just an exported member of the wallet, so it could be accessed directly without a getter.

manan19 replied:
74fccec47b (export Config instead of GetConfig())

jrick commented on wallet/wallet.go:
```diff
@@ -1041,7 +1047,7 @@ func (w *Wallet) ImportPrivateKey(wif *btcutil.WIF, bs *waddrmgr.BlockStamp,
 // ExportWatchingWallet returns a watching-only version of the wallet serialized
 // in a map.
-func (w *Wallet) ExportWatchingWallet() (map[string]string, error) {
+func (w *Wallet) ExportWatchingWallet(walletPass string) (map[string]string, error) {
```
The public passphrase should be reachable via the config, through the wallet receiver.  No need to pass it as a parameter.

manan19 replied:
addressed

jrick commented on config.go:
```diff
@@ -366,8 +358,17 @@ func loadConfig() (*config, []string, error) {
+		DisallowFree:			cfg.DisallowFree,
+		UseAddrIndex:			cfg.UseAddrIndex,
+		ChainParams:			activeNet.Params,
+		WalletDbName:			walletDbName,
```
The db filename shouldn't be a config option.  Instead, move the constant from the old config.go to the new package.

davecgh commented on wallet/wallet.go:
```diff
@@ -1305,13 +1311,13 @@ func (w *Wallet) TxRecord(txSha *wire.ShaHash) (r *txstore.TxRecord, ok bool) {
-	netdir := networkDir(cfg.DataDir, activeNet.Params)
-	dbPath := filepath.Join(netdir, walletDbName)
+// OpenWallet opens a wallet from disk.
+func OpenWallet(config *Config) (*Wallet, error) {
```
This can probably be renamed to Open.  So callers would do `w, err := wallet.Open(cfg)`.

manan19 replied:
b3d77f5fab (rename to remove api redundancy)

davecgh commented on wallet/config.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+const (
+	walletDbWatchingOnlyName = "wowallet.db"
```
I think this should be caller defined (part of the Config) like the normal database name.

davecgh commented on wallet/createtx.go:
```diff
@@ -164,11 +165,13 @@ func createTx(
 	feeIncrement btcutil.Amount,
 	mgr *waddrmgr.Manager,
-	changeAddress func(*waddrmgr.BlockStamp) (btcutil.Address, error)) (
+	changeAddress func(*waddrmgr.BlockStamp) (btcutil.Address, error),
```
Not sure what happened here with this function having one parameter per line, but this doesn't match the style of the rest of the code.  I realized you just followed suit, but now that I'm reviewing this and see it, if you would, please go ahead and fix the other params too so it's consistent with the rest of the code base.

manan19 replied:
a6202e166e (go fmt. reformat function declaration.)

davecgh commented on wallet/disksync.go:
```diff
@@ -23,7 +23,7 @@ import (
 // checkCreateDir checks that the path exists and is a directory.
 // If path does not exist, it is created.
-func checkCreateDir(path string) error {
+func CheckCreateDir(path string) error {
```
I not sure if this should be exported from the package.  It's basic operations that I'm not really sure make much sense to be exposed from this package.

manan19 replied:
fd1ab5c0a1 (basic util should not be exported from package)

davecgh commented on wallet/wallet.go:
```diff
@@ -60,7 +65,7 @@ const (
 // networkDir returns the directory name of a network directory to hold wallet
 // files.
-func networkDir(dataDir string, chainParams *chaincfg.Params) string {
+func NetworkDir(dataDir string, chainParams *chaincfg.Params) string {
```
I'd think I'd prefer the call is responsible for locating the wallet database.

davecgh commented:
In general, I'd like to see everything dealing with the creation of the wallet database and where it lives completely moved out of the new package.

It would be much cleaner to leave it up to the caller where to locate the database file, when to create a new one, etc.  The wallet package, imo, should just take a `walletdb.Db` that has already been opened.

davecgh commented on wallet/wallet.go:
```diff
@@ -1305,13 +1311,13 @@ func (w *Wallet) TxRecord(txSha *wire.ShaHash) (r *txstore.TxRecord, ok bool) {
-	dbPath := filepath.Join(netdir, walletDbName)
+// OpenWallet opens a wallet from disk.
+func OpenWallet(config *Config) (*Wallet, error) {
+	netdir := NetworkDir(config.DataDir, config.ChainParams)
```
As mentioned elsewhere, I'd prefer all of this be up to the caller and this function takes an already opened `walletdb.DB` to work with.

jrick commented:
I don't think passing just an opened DB to the wallet package is enough at the moment, since txstore is outside the db.

But we could pass both in, and then when wtxmgr goes live, introduce a breaking API change which removes the txstore option.

davecgh commented:
Agreed for btcwallet.  By making the caller responsible for the db, that is a choice we can make for btcwallet.  On the other hand, perhaps GUIs might want to support multiple wallets in the same directory, one wallet in multiple directories, or some combination thereof.  Making the caller responsible for such things offers the flexibility to do all of these things.  Having the wallet responsible for it like it is now doesn't.

manan19 commented:
Moved out wallet files creation as well as ownership outside the wallet package. The wallet database and txStore need to be passed in as parameters to wallet.Open

davecgh commented:
Status Update: We're currently working to get #155 and #204 merged to master.  This will need to be rebased and updated to handle those changes once they're merged.  After that we'll give this another round of review.

davecgh commented:
Status Update: #155 has been merged to master.  We're working on #204 now.  Once that is one, we'll work on this one.

davecgh commented on wallet/README.md:
```diff
@@ -0,0 +1,31 @@
+wallet
+========
```
Too many `=`.  The other READMEs make these match.

davecgh commented on wallet/README.md:
```diff
@@ -0,0 +1,31 @@
+
+## Feature Overview
+
+
```
Let's flesh this section out.

jrick replied:
The main feature of this package is the integration of all the other wallet components (waddrmgr, txstore, and chain) and the Wallet type is 'runnable', so it will be continuously updating itself against changes notified by the remote btcd instance.

davecgh commented on config.go:
```diff
@@ -24,24 +24,33 @@ import (
+
+	// maxEmptyAccounts is the number of accounts to scan even if they have no
+	// transaction history. This is a deviation from BIP044 to make account
+	// creation more easier by allowing a limited number of empty accounts.
```
more easier -> easier

davecgh commented on wallet/disksync.go:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2013, 2014 Conformal Systems LLC <info@conformal.com>
```
New file -> 2015

davecgh commented on wallet/wallet.go:
```diff
@@ -1437,60 +1369,15 @@ func (w *Wallet) TxRecord(txSha *wire.ShaHash) (r *txstore.TxRecord, ok bool) {
-			return nil, err
-		}
+// Returns wallet db being used by a wallet
+func (w *Wallet) GetDb() walletdb.DB {
```
Please name this `.Db()` per [Effective Go Guidelines](http://golang.org/doc/effective_go.html#Getters).

davecgh commented on walletsetup.go:
```diff
@@ -484,3 +557,88 @@ func createSimulationWallet(cfg *config) error {
 }
+
+func openDb(directory string, dbname string) (
+	*walletdb.DB, error) {
```
This should be on the line before it.

davecgh commented on walletsetup.go:
```diff
@@ -484,3 +557,88 @@ func createSimulationWallet(cfg *config) error {
 	return nil
 }
+
+func openDb(directory string, dbname string) (
```
Please comment the intent, caveats, etc of this function.

davecgh commented on walletsetup.go:
```diff
@@ -484,3 +557,88 @@ func createSimulationWallet(cfg *config) error {
+	return &db, nil
+}
+
+func openWaddrmgr(db *walletdb.DB, namespaceKey []byte, pass string,
```
Please comment the intent, caveats, etc of this function.

davecgh commented on walletsetup.go:
```diff
@@ -484,3 +557,88 @@ func createSimulationWallet(cfg *config) error {
+		chainParams, config)
+}
+
+func openWallet() (*wallet.Wallet, error) {
```
Please comment the intent, caveats, etc of this function.

davecgh commented on log.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/btcsuite/btcwallet/chain"
 	"github.com/btcsuite/btcwallet/txstore"
 	"github.com/btcsuite/seelog"
+	"github.com/mirrorx/btcwallet/wallet"
```
btcsuite

davecgh commented on walletsetup.go:
```diff
@@ -26,15 +26,88 @@ import (
 	"github.com/btcsuite/btcwallet/walletdb"
 	_ "github.com/btcsuite/btcwallet/walletdb/bdb"
 	"github.com/btcsuite/golangcrypto/ssh/terminal"
+	"github.com/mirrorx/btcwallet/wallet"
```
btcsuite

davecgh commented on rpcserver.go:
```diff
@@ -48,6 +48,7 @@ import (
 	"github.com/btcsuite/btcwallet/txstore"
 	"github.com/btcsuite/btcwallet/waddrmgr"
 	"github.com/btcsuite/websocket"
+	"github.com/mirrorx/btcwallet/wallet"
```
btcsuite

davecgh commented:
Also, please create a `doc.go` file along the lines of https://github.com/btcsuite/btcwallet/blob/master/waddrmgr/doc.go.

davecgh commented:
All changes reviewed.  As discussed in IRC, please add TODOs in `doc.go` and `README.md` for the documentation to be completed by #224, and squash/rebase this to the latest master.  Then I'll be able to OK it and get it merged.

manan19 commented:
@davecgh 
- Added 2 TODO's in wallet/doc.go and wallet/README.md. 
- Rebased, squashed into one descriptive commit and force pushed.

Travis build seems good, ready to be merged in.
